Member functions as Middleware*

Do somebody ever tried to build a library over this server?
Since "ESPAsyncWiFiManager" is not working over Ethernet, I am trying to build my library for a Webmanager which works over Ethernet, but I can´t figure out how to pass member functions of my EthWebManager class as middlewares.

I am trying with std::bind, here is a code snippet:

void EthWebManager::auth(Request &req, Response &res) {
  char * authHeader = req.get("Authorization");

  if (!str_iseq(authHeader, encoded_login1.c_str()) && !str_iseq(authHeader, encoded_login2.c_str())) {
    res.set("WWW-Authenticate", "Basic realm=\"Secret Area\"");
    res.sendStatus(401);
    res.end();
    DEBUG_WM(F("Unauthenticated"));
  }
}

void EthWebManager::handleRoot(Request &req, Response &res)
{
  String page = FPSTR(WFM_HTTP_HEAD);
  page.replace("{v}", "Home");
  page += FPSTR(HTTP_SCRIPT);
  page += FPSTR(HTTP_STYLE);
  page += FPSTR(HTTP_HEAD_END);
  page += F("<h3>EthWebManager</h3>");
  page += FPSTR(HTTP_PORTAL_OPTIONS);
  page += FPSTR(HTTP_END);
  res.set("Content-Type", "text/html");
  res.print(page);
}

void EthWebManager::startEthManager()
{
  server->begin();
  delay(500);
  mainapp.header("Authorization", authBuffer, 200);
  mainapp.use(std::bind(&EthWebManager::auth, this, std::placeholders::_1, std::placeholders::_2));
  auto callable = std::bind(&EthWebManager::handleRoot, this, std::placeholders::_1, std::placeholders::_2);
  mainapp.get("/", callable );
}


**But I keep getting these errors:**

lib\EthWebManager\src\EthWebManager.cpp: In member function 'void EthWebManager::startEthManager()':
lib\EthWebManager\src\EthWebManager.cpp:361:98: error: no matching function for call to 'awot::Application::use(std::_Bind_helper<false, void (EthWebManager::*)(awot::Request&, awot::Response&), EthWebManager*, const std::_Placeholder<1>&, const std::_Placeholder<2>&>::type)'
   mainapp.use(std::bind(&EthWebManager::auth, this, std::placeholders::_1, std::placeholders::_2));
                                                                                                  ^
In file included from lib\EthWebManager\src\EthWebManager.h:22:0,
                 from lib\EthWebManager\src\EthWebManager.cpp:14:
.pio\libdeps\m5stack-core-esp32\aWOT\src/aWOT.h:322:8: note: candidate: void awot::Application::use(const char*, awot::Router*)
   void use(const char* path, Router* router);
        ^
.pio\libdeps\m5stack-core-esp32\aWOT\src/aWOT.h:322:8: note:   candidate expects 2 arguments, 1 provided
.pio\libdeps\m5stack-core-esp32\aWOT\src/aWOT.h:323:8: note: candidate: void awot::Application::use(awot::Router*)
   void use(Router* router);
        ^
.pio\libdeps\m5stack-core-esp32\aWOT\src/aWOT.h:323:8: note:   no known conversion for argument 1 from 'std::_Bind_helper<false, void (EthWebManager::*)(awot::Request&, awot::Response&), EthWebManager*, const std::_Placeholder<1>&, const std::_Placeholder<2>&>::type {aka std::_Bind<std::_Mem_fn<void (EthWebManager::*)(awot::Request&, awot::Response&)>(EthWebManager*, std::_Placeholder<1>, std::_Placeholder<2>)>}' to 'awot::Router*'
.pio\libdeps\m5stack-core-esp32\aWOT\src/aWOT.h:324:8: note: candidate: void awot::Application::use(const char*, void (*)(awot::Request&, awot::Response&))
   void use(const char* path, Router::Middleware* middleware);
        ^
.pio\libdeps\m5stack-core-esp32\aWOT\src/aWOT.h:324:8: note:   candidate expects 2 arguments, 1 provided
.pio\libdeps\m5stack-core-esp32\aWOT\src/aWOT.h:325:8: note: candidate: void awot::Application::use(void (*)(awot::Request&, awot::Response&))
   void use(Router::Middleware* middleware);
        ^
.pio\libdeps\m5stack-core-esp32\aWOT\src/aWOT.h:325:8: note:   no known conversion for argument 1 from 'std::_Bind_helper<false, void (EthWebManager::*)(awot::Request&, awot::Response&), EthWebManager*, const std::_Placeholder<1>&, const std::_Placeholder<2>&>::type {aka std::_Bind<std::_Mem_fn<void (EthWebManager::*)(awot::Request&, awot::Response&)>(EthWebManager*, std::_Placeholder<1>, std::_Placeholder<2>)>}' to 'void (*)(awot::Request&, awot::Response&)'
lib\EthWebManager\src\EthWebManager.cpp:363:29: error: no matching function for call to 'awot::Application::get(const char [2], std::_Bind<std::_Mem_fn<void (EthWebManager::*)(awot::Request&, awot::Response&)>(EthWebManager*, std::_Placeholder<1>, std::_Placeholder<2>)>&)'
   mainapp.get("/", callable ); /*
                             ^
In file included from lib\EthWebManager\src\EthWebManager.h:22:0,
                 from lib\EthWebManager\src\EthWebManager.cpp:14:
.pio\libdeps\m5stack-core-esp32\aWOT\src/aWOT.h:300:8: note: candidate: void awot::Application::get(const char*, void (*)(awot::Request&, awot::Response&))
   void get(const char* path, Router::Middleware* middleware);
        ^
.pio\libdeps\m5stack-core-esp32\aWOT\src/aWOT.h:300:8: note:   no known conversion for argument 2 from 'std::_Bind<std::_Mem_fn<void (EthWebManager::*)(awot::Request&, awot::Response&)>(EthWebManager*, std::_Placeholder<1>, std::_Placeholder<2>)>' to 'void (*)(awot::Request&, awot::Response&)'   
.pio\libdeps\m5stack-core-esp32\aWOT\src/aWOT.h:301:8: note: candidate: void awot::Application::get(void (*)(awot::Request&, awot::Response&))
   void get(Router::Middleware* middleware);
        ^
.pio\libdeps\m5stack-core-esp32\aWOT\src/aWOT.h:301:8: note:   candidate expects 1 argument, 2 provided
*** [.pio\build\m5stack-core-esp32\lib4c6\EthWebManager\EthWebManager.cpp.o] Error 1

--- a/examples/MemberFunction/.arduino-ci.yml
+++ b/examples/MemberFunction/.arduino-ci.yml
@@ -1,0 +1,3 @@
+compile:
+  platforms:
+    - esp32

--- a/examples/MemberFunction/.arduino-ci.yml
+++ b/examples/MemberFunction/.arduino-ci.yml
@@ -1,3 +1,2 @@
 compile:
   platforms:
-    - esp32

--- a/examples/MemberFunction/MemberFunction.ino
+++ b/examples/MemberFunction/MemberFunction.ino
@@ -1,0 +1,64 @@
+#if defined(ESP8266)
+#include <ESP8266WiFi.h>
+#else
+#include <WiFi.h>
+#endif
+
+#define STD_FUNCTION_MIDDLEWARE
+#include "aWOT.h"
+
+#define WIFI_SSID ""
+#define WIFI_PASSWORD ""
+
+WiFiServer server(80);
+Application app;
+
+class Counter {
+  public:
+    Counter(int initialCount) {
+      _count = initialCount;
+    }
+
+    void handleRequest(Request &req, Response &res) {
+      _count++;
+      res.print("I have been called ");
+      res.print(_count);
+      res.print(" times!");
+    }
+
+  private:
+    int _count = 0;
+};
+
+Counter counter(666);
+
+void callback(Request &req, Response &res) {
+  res.print("Hello World!");
+}
+
+void setup() {
+  Serial.begin(115200);
+
+  WiFi.begin(WIFI_SSID, WIFI_PASSWORD);
+  while (WiFi.status() != WL_CONNECTED) {
+    delay(500);
+    Serial.print(".");
+  }
+  Serial.println(WiFi.localIP());
+
+
+  app.get("/", std::bind(&Counter::handleRequest, &counter, std::placeholders::_1, std::placeholders::_2));
+
+  // Do not add '&' in front of the parameter when STD_FUNCTION_MIDDLEWARE has been defined.
+  app.get("/callback", callback);
+  server.begin();
+}
+
+void loop() {
+  WiFiClient client = server.available();
+
+  if (client.connected()) {
+    app.process(&client);
+    client.stop();
+  }
+}

--- a/examples/MemberFunction/MemberFunction.ino
+++ b/examples/MemberFunction/MemberFunction.ino
@@ -4,8 +4,8 @@
 #include <WiFi.h>
 #endif
 
-#define STD_FUNCTION_MIDDLEWARE
-#include "aWOT.h"
+// Manually add "#define STD_FUNCTION_MIDDLEWARE" to the top of the aWOT.h file
+#include <aWOT.h>
 
 #define WIFI_SSID ""
 #define WIFI_PASSWORD ""

--- a/library.json
+++ b/library.json
@@ -1,6 +1,6 @@
 {
   "name": "aWOT",
-  "version": "3.4.0",
+  "version": "3.5.0",
   "keywords": "wifi, ethernet, http, web, server, rest",
   "description": "aWOT is a web server library compatible with multiple different board architectures and networking solutions.",
   "license": "MIT",

--- a/library.properties
+++ b/library.properties
@@ -1,5 +1,5 @@
 name=aWOT
-version=3.4.0
+version=3.5.0
 author=Lasse Lukkari <lasse.lukkari@gmail.com>
 maintainer=Lasse Lukkari <lasse.lukkari@gmail.com>
 sentence=Arduino web server library.

--- a/src/aWOT.cpp
+++ b/src/aWOT.cpp
@@ -1283,67 +1283,67 @@ Router::~Router() {
   m_head = NULL;
 }
 
-void Router::del(const char *path, Middleware *middleware) {
+void Router::del(const char *path, MIDDLEWARE_PARAM middleware) {
   m_addMiddleware(Request::DELETE, path, middleware);
 }
 
-void Router::del(Middleware *middleware) {
+void Router::del(MIDDLEWARE_PARAM middleware) {
   del(NULL, middleware);
 }
 
-void Router::get(const char *path, Middleware *middleware) {
+void Router::get(const char *path, MIDDLEWARE_PARAM middleware) {
   m_addMiddleware(Request::GET, path, middleware);
 }
 
-void Router::get(Middleware *middleware) {
+void Router::get(MIDDLEWARE_PARAM middleware) {
   get(NULL, middleware);
 }
 
-void Router::head(const char *path, Middleware *middleware) {
+void Router::head(const char *path, MIDDLEWARE_PARAM middleware) {
   m_addMiddleware(Request::HEAD, path, middleware);
 }
 
-void Router::head(Middleware *middleware) {
+void Router::head(MIDDLEWARE_PARAM middleware) {
   head(NULL, middleware);
 }
 
-void Router::options(const char *path, Middleware *middleware) {
+void Router::options(const char *path, MIDDLEWARE_PARAM middleware) {
   m_addMiddleware(Request::OPTIONS, path, middleware);
 }
 
-void Router::options(Middleware *middleware) {
+void Router::options(MIDDLEWARE_PARAM middleware) {
   options(NULL, middleware);
 }
 
-void Router::post(const char *path, Middleware *middleware) {
+void Router::post(const char *path, MIDDLEWARE_PARAM middleware) {
   m_addMiddleware(Request::POST, path, middleware);
 }
 
-void Router::post(Middleware *middleware) {
+void Router::post(MIDDLEWARE_PARAM middleware) {
   post(NULL, middleware);
 }
 
-void Router::put(const char *path, Middleware *middleware) {
+void Router::put(const char *path, MIDDLEWARE_PARAM middleware) {
   m_addMiddleware(Request::PUT, path, middleware);
 }
 
-void Router::put(Middleware *middleware) {
+void Router::put(MIDDLEWARE_PARAM middleware) {
   put(NULL, middleware);
 }
 
-void Router::patch(const char *path, Middleware *middleware) {
+void Router::patch(const char *path, MIDDLEWARE_PARAM middleware) {
   m_addMiddleware(Request::PATCH, path, middleware);
 }
 
-void Router::patch(Middleware *middleware) {
+void Router::patch(MIDDLEWARE_PARAM middleware) {
   patch(NULL, middleware);
 }
 
-void Router::use(const char *path, Middleware *middleware) {
+void Router::use(const char *path, MIDDLEWARE_PARAM middleware) {
   m_addMiddleware(Request::ALL, path, middleware);
 }
 
-void Router::use(Middleware *middleware) {
+void Router::use(MIDDLEWARE_PARAM middleware) {
   use(NULL, middleware);
 }
 
@@ -1361,7 +1361,7 @@ void Router::use(Router *router) {
 }
 
 void Router::m_addMiddleware(Request::MethodType type, const char *path,
-                             Middleware *middleware) {
+                             MIDDLEWARE_PARAM middleware) {
   MiddlewareNode *tail = new MiddlewareNode();
   tail->path = path;
   tail->middleware = middleware;
@@ -1507,67 +1507,67 @@ Application::~Application() {
   m_headerTail = NULL;
 }
 
-void Application::del(const char *path, Router::Middleware *middleware) {
+void Application::del(const char *path, Router::MIDDLEWARE_PARAM middleware) {
   m_defaultRouter.m_addMiddleware(Request::DELETE, path, middleware);
 }
 
-void Application::del(Router::Middleware *middleware) {
+void Application::del(Router::MIDDLEWARE_PARAM middleware) {
   del(NULL, middleware);
 }
 
-void Application::finally(Router::Middleware *final) {
+void Application::finally(Router::MIDDLEWARE_PARAM final) {
   m_final = final;
 }
 
-void Application::get(const char *path, Router::Middleware *middleware) {
+void Application::get(const char *path, Router::MIDDLEWARE_PARAM middleware) {
   m_defaultRouter.m_addMiddleware(Request::GET, path, middleware);
 }
 
-void Application::get(Router::Middleware *middleware) {
+void Application::get(Router::MIDDLEWARE_PARAM middleware) {
   get(NULL, middleware);
 }
 
-void Application::head(const char *path, Router::Middleware *middleware) {
+void Application::head(const char *path, Router::MIDDLEWARE_PARAM middleware) {
   m_defaultRouter.m_addMiddleware(Request::HEAD, path, middleware);
 }
 
-void Application::head(Router::Middleware *middleware) {
+void Application::head(Router::MIDDLEWARE_PARAM middleware) {
   head(NULL, middleware);
 }
 
-void Application::notFound(Router::Middleware *notFound) {
+void Application::notFound(Router::MIDDLEWARE_PARAM notFound) {
   m_notFound = notFound;
 }
 
-void Application::options(const char *path, Router::Middleware *middleware) {
+void Application::options(const char *path, Router::MIDDLEWARE_PARAM middleware) {
   m_defaultRouter.m_addMiddleware(Request::OPTIONS, path, middleware);
 }
 
-void Application::options(Router::Middleware *middleware) {
+void Application::options(Router::MIDDLEWARE_PARAM middleware) {
   options(NULL, middleware);
 }
 
-void Application::patch(const char *path, Router::Middleware *middleware) {
+void Application::patch(const char *path, Router::MIDDLEWARE_PARAM middleware) {
   m_defaultRouter.m_addMiddleware(Request::PATCH, path, middleware);
 }
 
-void Application::patch(Router::Middleware *middleware) {
+void Application::patch(Router::MIDDLEWARE_PARAM middleware) {
   patch(NULL, middleware);
 }
 
-void Application::post(const char *path, Router::Middleware *middleware) {
+void Application::post(const char *path, Router::MIDDLEWARE_PARAM middleware) {
   m_defaultRouter.m_addMiddleware(Request::POST, path, middleware);
 }
 
-void Application::post(Router::Middleware *middleware) {
+void Application::post(Router::MIDDLEWARE_PARAM middleware) {
   post(NULL, middleware);
 }
 
-void Application::put(const char *path, Router::Middleware *middleware) {
+void Application::put(const char *path, Router::MIDDLEWARE_PARAM middleware) {
   m_defaultRouter.m_addMiddleware(Request::PUT, path, middleware);
 }
 
-void Application::put(Router::Middleware *middleware) {
+void Application::put(Router::MIDDLEWARE_PARAM middleware) {
   put(NULL, middleware);
 }
 
@@ -1641,11 +1641,11 @@ void Application::process(Stream *stream, char *urlBuffer, int urlBufferLength, 
   process(&client, urlBuffer, urlBufferLength, writeBuffer, writeBufferLength, context);
 }
 
-void Application::use(const char *path, Router::Middleware *middleware) {
+void Application::use(const char *path, Router::MIDDLEWARE_PARAM middleware) {
   m_defaultRouter.m_addMiddleware(Request::ALL, path, middleware);
 }
 
-void Application::use(Router::Middleware *middleware) {
+void Application::use(Router::MIDDLEWARE_PARAM middleware) {
   use(NULL, middleware);
 }
 

--- a/src/aWOT.cpp
+++ b/src/aWOT.cpp
@@ -33,7 +33,6 @@ Response::Response(Client* client, uint8_t * writeBuffer, int writeBufferLength)
       m_sendingStatus(false),
       m_sendingHeaders(false),
       m_headersCount(0),
-      m_mime(NULL),
       m_bytesSent(0),
       m_ended(false),
       m_buffer(writeBuffer),

--- a/src/aWOT.h
+++ b/src/aWOT.h
@@ -163,7 +163,6 @@ class Response : public Print {
   bool m_sendingStatus;
   bool m_sendingHeaders;
   int m_headersCount;
-  char* m_mime;
   int m_bytesSent;
   bool m_ended;
   uint8_t * m_buffer;

--- a/src/aWOT.h
+++ b/src/aWOT.h
@@ -29,6 +29,15 @@
 
 #include "Client.h"
 
+#if defined(STD_FUNCTION_MIDDLEWARE)
+#include <functional>
+#define MIDDLEWARE_PARAM  Middleware
+#define MIDDLEWARE_FUNCTION std::function<void(Request& request, Response& response)> Middleware
+#else
+#define MIDDLEWARE_PARAM  Middleware*
+#define MIDDLEWARE_FUNCTION void Middleware(Request& request, Response& response)
+#endif
+
 #define CRLF "\r\n"
 
 #if defined(__AVR_ATmega328P__) || defined(__AVR_Atmega32U4__) || \
@@ -242,41 +251,41 @@ class Router {
   friend class Application;
 
  public:
-  typedef void Middleware(Request& request, Response& response);
+  typedef MIDDLEWARE_FUNCTION;
 
   Router();
   ~Router();
 
-  void del(const char* path, Middleware* middleware);
-  void del(Middleware* middleware);
-  void get(const char* path, Middleware* middleware);
-  void get(Middleware* middleware);
-  void head(const char* path, Middleware* middleware);
-  void head(Middleware* middleware);
-  void options(const char* path, Middleware* middleware);
-  void options(Middleware* middleware);
-  void patch(const char* path, Middleware* middleware);
-  void patch(Middleware* middleware);
-  void post(const char* path, Middleware* middleware);
-  void post(Middleware* middleware);
-  void put(const char* path, Middleware* middleware);
-  void put(Middleware* middleware);
+  void del(const char* path, MIDDLEWARE_PARAM middleware);
+  void del(MIDDLEWARE_PARAM middleware);
+  void get(const char* path, MIDDLEWARE_PARAM middleware);
+  void get(MIDDLEWARE_PARAM middleware);
+  void head(const char* path, MIDDLEWARE_PARAM middleware);
+  void head(MIDDLEWARE_PARAM middleware);
+  void options(const char* path, MIDDLEWARE_PARAM middleware);
+  void options(MIDDLEWARE_PARAM middleware);
+  void patch(const char* path, MIDDLEWARE_PARAM middleware);
+  void patch(MIDDLEWARE_PARAM middleware);
+  void post(const char* path, MIDDLEWARE_PARAM middleware);
+  void post(MIDDLEWARE_PARAM middleware);
+  void put(const char* path, MIDDLEWARE_PARAM middleware);
+  void put(MIDDLEWARE_PARAM middleware);
   void use(const char* path, Router* router);
   void use(Router* router);
-  void use(const char* path, Middleware* middleware);
-  void use(Middleware* middleware);
+  void use(const char* path, MIDDLEWARE_PARAM middleware);
+  void use(MIDDLEWARE_PARAM middleware);
 
  private:
   struct MiddlewareNode {
     const char* path;
-    Middleware* middleware;
+    MIDDLEWARE_PARAM middleware;
     Router* router;
     Request::MethodType type;
     MiddlewareNode* next;
   };
 
   void m_addMiddleware(Request::MethodType type, const char* path,
-                       Middleware* middleware);
+                       MIDDLEWARE_PARAM middleware);
   void m_mountMiddleware(MiddlewareNode *tail);
   void m_setNext(Router* next);
   Router* m_getNext();
@@ -294,23 +303,23 @@ class Application {
   static int strcmpi(const char* s1, const char* s2);
   static int strcmpiP(const char* s1, const unsigned char* s2);
 
-  void del(const char* path, Router::Middleware* middleware);
-  void del(Router::Middleware* middleware);
-  void finally(Router::Middleware* middleware);
-  void get(const char* path, Router::Middleware* middleware);
-  void get(Router::Middleware* middleware);
-  void head(const char* path, Router::Middleware* middleware);
-  void head(Router::Middleware* middleware);
+  void del(const char* path, Router::MIDDLEWARE_PARAM middleware);
+  void del(Router::MIDDLEWARE_PARAM middleware);
+  void finally(Router::MIDDLEWARE_PARAM middleware);
+  void get(const char* path, Router::MIDDLEWARE_PARAM middleware);
+  void get(Router::MIDDLEWARE_PARAM middleware);
+  void head(const char* path, Router::MIDDLEWARE_PARAM middleware);
+  void head(Router::MIDDLEWARE_PARAM middleware);
   void header(const char* name, char* buffer, int bufferLength);
-  void notFound(Router::Middleware* middleware);
-  void options(const char* path, Router::Middleware* middleware);
-  void options(Router::Middleware* middleware);
-  void patch(const char* path, Router::Middleware* middleware);
-  void patch(Router::Middleware* middleware);
-  void post(const char* path, Router::Middleware* middleware);
-  void post(Router::Middleware* middleware);
-  void put(const char* path, Router::Middleware* middleware);
-  void put(Router::Middleware* middleware);
+  void notFound(Router::MIDDLEWARE_PARAM middleware);
+  void options(const char* path, Router::MIDDLEWARE_PARAM middleware);
+  void options(Router::MIDDLEWARE_PARAM middleware);
+  void patch(const char* path, Router::MIDDLEWARE_PARAM middleware);
+  void patch(Router::MIDDLEWARE_PARAM middleware);
+  void post(const char* path, Router::MIDDLEWARE_PARAM middleware);
+  void post(Router::MIDDLEWARE_PARAM middleware);
+  void put(const char* path, Router::MIDDLEWARE_PARAM middleware);
+  void put(Router::MIDDLEWARE_PARAM middleware);
   void process(Client* client, void* context = NULL);
   void process(Client* client, char* urlbuffer, int urlBufferLength, void* context = NULL);
   void process(Client* client, char* urlBuffer, int urlBufferLength, uint8_t * writeBuffer, int writeBufferLength, void* context = NULL);
@@ -321,14 +330,14 @@ class Application {
   void setTimeout(unsigned long timeoutMillis);
   void use(const char* path, Router* router);
   void use(Router* router);
-  void use(const char* path, Router::Middleware* middleware);
-  void use(Router::Middleware* middleware);
+  void use(const char* path, Router::MIDDLEWARE_PARAM middleware);
+  void use(Router::MIDDLEWARE_PARAM middleware);
 
  private:
   void m_process(Request &req, Response &res);
 
-  Router::Middleware* m_final;
-  Router::Middleware* m_notFound;
+  Router::MIDDLEWARE_PARAM m_final;
+  Router::MIDDLEWARE_PARAM m_notFound;
   Router m_defaultRouter;
   Request::HeaderNode* m_headerTail;
   unsigned long m_timeout;

